internal/codeintel/uploads/internal/background/processor/job_worker_handler.go

```diff
@@ -402,13 +432,67 @@ func withUploadData(ctx context.Context, logger log.Logger, uploadStore uploadst
 	}
 	defer rc.Close()
 
-	rc, err = gzip.NewReader(rc)
-	if err != nil {
-		return errors.Wrap(err, "gzip.NewReader")
+	trySaveToDisk := func(_ io.Reader) (gzipReadSeeker, *os.File, error) {
+		var indexReader gzipReadSeeker
+		tmpFilePattern := fmt.Sprintf("upload-%d-tmp.gz", uploadStats.ID)
+		tmpFile, err := os.CreateTemp("", tmpFilePattern)
+		if err != nil {
+			logger.Warn("Failed to create temporary file to save upload for streaming",
+				log.NamedError("err", err),
+				log.String("filename", uploadFilename))
+		} else if indexReader, err = newGzipReadSeeker(tmpFile); err != nil {
+			return indexReader, nil, errors.Wrapf(err, "failed to seek for file: %s", tmpFile.Name())
+		}
+		return indexReader, tmpFile, nil
+	}
+
+	var readerForDiskWrite io.Reader = nil
+	var buf []byte
+	compressedSizeHint := int64(0)
+	if uploadStats.UploadSize != nil {
+		compressedSizeHint = *uploadStats.UploadSize
+	}
+	if uploadStats.UncompressedSize != nil && *uploadStats.UncompressedSize > uncompressedSizeLimitBytes {
+		readerForDiskWrite = rc
+	} else {
+		buf, err = readAllWithSizeHint(rc, compressedSizeHint)
+		if err != nil {
+			return errors.Wrap(err, "failed to read upload file")
+		}
+		compressedSize := len(buf)
+		// The factor of 5 is based on ~worst-case gzip compression ratio.
+		// See NOTE(scip-index-size-stats)
+		uncompressedSizeEstimate := compressedSize * 5
+		if uncompressedSizeEstimate > uncompressedSizeLimitBytes {
+			readerForDiskWrite = bytes.NewReader(buf)
+		}
+	}
+	var indexReader gzipReadSeeker
+	if readerForDiskWrite != nil {
+		var tmpFile *os.File
+		indexReader, tmpFile, err = trySaveToDisk(readerForDiskWrite)
+		if tmpFile != nil {
+			defer os.RemoveAll(tmpFile.Name())
+		}
+		if err != nil {
+			return err
+		}
```

Comment:
Would be good if we could have `trySaveToDisk` clean up after itself on failure.

```suggestion
		if err != nil {
			return err
		}
		defer os.RemoveAll(tmpFile.Name())
```